Sources/Conduit/Auth/OAuth2RequestPipelineMiddleware.swift

```diff
@@ -54,8 +59,22 @@ public struct OAuth2RequestPipelineMiddleware: RequestPipelineMiddleware {
         else if let token = token,
             token.refreshToken != nil,
             refreshStrategyFactory != nil {
+            if tokenStorage.isRefreshTokenLockedFor(client: clientConfiguration, authorization: authorization),
+                let tokenLockExpiration = tokenStorage.refreshTokenLockExpirationFor(client: clientConfiguration, authorization: authorization) {
+                logger.info("Token refresh is active in an alternate session; retrying once lock is relinquished")
+                let timeout = max(0, tokenLockExpiration.timeIntervalSinceNow)
+                OAuth2TokenRefreshCoordinator.shared.waitForRefresh(timeout: timeout) {
+                    /// If we hit the edge case where an unrelated session triggers OAuth2TokenRefreshCoordinator.endTokenRefresh(),
+                    /// then we'll recursively fall back into this branch
+                    self.prepareForTransport(request: request, completion: completion)
+                }
+                return
+            }
+            tokenStorage.lockRefreshToken(timeout: tokenRefreshLockRelinquishInterval, client: clientConfiguration, authorization: authorization)
```

Comment:
This two methods are not thread-safe, and are prone to race conditions. This is, two processes can call `isRefreshTokenLockedFor` and get `false`, and then both hit the `lockRefreshToken` method before the lock is in place.

Is there a way we can remove the `isRefreshTokenLockedFor` check? Ideally, we would want processes to hit the lock and wait for the unlock, in the same way `NSLock` or semaphores work.

Reply:
Responded in below comment